larq/quantizers.py

```diff
@@ -601,9 +604,95 @@ def get_config(self):
         return {**super().get_config(), "k_bit": self.precision}
 
 
+@utils.register_alias("dorefa_kernel_quantizer")
+@utils.register_keras_custom_object
+class DoReFaKernel(DoReFa):
```

Comment:
I chose this behavior, because a new class integrates nicely into the automatic creation of the docs and into the unittest framework, where everything is optimized for "one lass per quantizer behavior". I thought about using an instance of the activation quantizer inside the weight quantizer instead of using inheritance, but I am not sure about side effets, I would need to forward arguments to this inner instance and I would need to forward calls to get_config. Using inheritance was the cleanest solution I found.

Reply:
For cleanliness of the code I'd prefer to have just a single quantizer class (e.g. like the preliminary version I wrote [here](https://github.com/larq/larq/issues/487#issuecomment-630748936)), but the docs are a good point. I think I would slightly prefer having a base `DoReFa` with `mode="activations"` and then just create a `DoReFaKernel` wrapper that sets `mode="weights"`. No strong opinions though.

Reply:
This decision affects building the docs and testing, where I do not know the larq project as good as others do. Please agree on one approach and I'll change (or not change) the code accordingly. I took the subclass-way, because it leaves a clear border between activation and weight quantizer without the need for forwarding any calls. Instead using a new keyword for the activation quantizer and some kind of a wrapper for creating the weight quantizer skips usage of inheritance and possible pitfalls.

Reply:
Thinking about this again, we can easily change the docs generation to add autogenerated plots also for quantizers with specific arguments (I am happy to take this on in a follow up PR since it requires changes in some not well documented code paths).

Following the argument from @jneeven above I think I now agree that using a single `DoReFa` class with a `mode` argument that defaults to `"activations"` is probably the more consistent approach within our current API. This might come up again in the future where different quantizers might need different behaviour depending on whether they are used for weights or activations. An example for this would be  int8 quantizers that require different per-channel setting depending on whether they are used for weights or activations. Adding separate classes for each of these would probably increase the API surface too much in my opinion.
I think I am also against exposing a separate `DoReFaKernel` wrapper in this case so this PR would reduce to adding support for a `mode` argument to the existing `DoReFa` class. Would this work for your use case?

Sorry for the back and forth but since this changes the public API I think we want to get it right so we don't have to make any breaking changes later on.

Reply:
This sounds great!